Set default suffix in "Save as..." file save dialog

zonkmachine opened the pull request:
When you "save as..." without adding suffix manually, LMMS will add a suffix but not test if the file "name + suffix" exists and will gladly write over an existing file.
This is true on Ubuntu Mate at least.
This patch adds suffix mmp/mmpz as default to VersionedSaveDialog when called in saveProjectAs() which fixes the issue.
When I searched the issue I got the impression that this could be mostly a Linux thing but I have not possibility to test on any other OS.

zonkmachine commented:
![saveas](https://cloud.githubusercontent.com/assets/6368949/9026107/b92185bc-3922-11e5-97e1-48475e8870bf.png)

Save action after fix.

zonkmachine commented:
After a cup of sufficiently strong coffee I realise that this should be fixed on the other save dialogues too so I'll expand on this pull request.

curlymorphic commented:
@zonkmachine Nice work, and good catch on the other save dialogs

Are you able to test windows builds using wine?

Hopefully we are not too far way from releasing test builds where full testing under windows will take place.

zonkmachine commented:
@curlymorphic Thanks!

> Are you able to test windows builds using wine?

No. It's a miracle that I manage to compile anything at all. Building under or for Wine are too many alien elements for me right now.

I've caught some of the cases now but I don't quite know what to do about the cases where the suffix is selected after the file dialogue calls exec()

zonkmachine commented:
To sum up the work so far.

:exclamation: Save As...		Fixed for .mmp and .mmpz but not for .mpt
:x: Export...		Not fixed.
:x: Export tracks...	Not looked at.
:heavy_check_mark: Export MIDI...	Fixed
:heavy_check_mark: Save preset		Fixed

zonkmachine commented:
Since we do the work for the .mpt outside of the sfd.exec() the default suffix is added so we need to remove it first. I'll go shopping for a better method but for now .mmp of .mmpz  works.

zonkmachine commented:
This should maybe be done in VersionedSaveDialog() but I can't wrap my grey matter around it. As an example I give you this instead.
( .mpt ) It works and the odds that you ever come across it is tiny. You need to first want to save a template file and also want to write over a present file or accidentally type the name of a present file... all this without typing in the suffix manually, then this message box will execute and since it's done after versionedSaveDialog() has finished it won't take you back to the save dialog but to the project window.

tresf commented:
@zonkmachine,

Here's an old commit where I tried to tackle a somewhat similar problem https://github.com/LMMS/lmms/commit/7a03353fb6b6d3f9091ae32d3b9e1d90f093271e (yeah, it's ugly)

zonkmachine commented:
> tackle a somewhat similar problem

As I understand it that part is for avoiding adding the suffix if the user has already typed it in. I basically just tossed a message box in there after the output file name is done to test for duplicates.
I'm uncertain of how this works on other platforms and I'm curious to see if there is a better way to do it within the file dialogue exec() and not after.

tresf commented:
> I'm uncertain of how this works on other platforms

Since we're using Qt's dialog, it should be identical across platforms.

> I'm curious to see if there is a better way to do it within the file dialogue exec() and not after.

Agreed.

tresf commented:
Perhaps... http://stackoverflow.com/questions/9285575/how-can-i-get-a-qfiledialog-to-prompt-for-overwrite-in-qt-4

zonkmachine commented:
> Perhaps... http://stackoverflow.com/questions/9285575/how-can-i-get-a-qfiledialog-to-prompt-for-overwrite-in-qt-4

No. We set ```QFileDialog::AcceptSave``` already in versionedFileDialog, among the first things actually.



> > I'm curious to see if there is a better way to do it within the file dialogue exec() and not after.
> 
> Agreed.

Maybe handle it under https://github.com/LMMS/lmms/issues/1834 then and just settle for a more temporary fix for now?

tresf commented:
@zonkmachine if overwrite prompting is going to continue to be a problem moving forward, can we permanently patch in this prompt directly in the [`FileDialog`](https://github.com/LMMS/lmms/blob/master/src/gui/dialogs/FileDialog.cpp) class?

zonkmachine commented:
Looking into it.

zonkmachine commented:
> can we permanently patch in this prompt directly in the FileDialog class?

Nope. I can't wrap my head around that one.

zonkmachine commented:
@tresf I don't know how to proceed with this one. The PR still works and I've updated it but haven't done anything more to it.

> @zonkmachine if overwrite prompting is going to continue to be a problem moving forward, can we permanently patch in this prompt directly in the FileDialog class?

Add to #1834 and fix later?

tresf commented on src/gui/MainWindow.cpp:
```diff
@@ -956,13 +956,41 @@ bool MainWindow::saveProjectAs()
+				{
+					QMessageBox mb;
+					mb.setWindowTitle( tr( "Save project" ) );
+					mb.setText( fname +  tr( ".mpt already exists. "
```
Probably shouldn't translate the `.mpt` portion

tresf commented:
In short, yes, this technique is our only technique.  I find [people's answer on this](http://www.qtcentre.org/archive/index.php/t-20156.html) to be utter bullshit.  Native common dialogs handle these things all the time and Qt should too. `</endrant>`.

Anyway, yes, this is the right approach.  Only criticism I have is we should probably consolidate as much of this logic into a single, static helper function.  `FileDialog` is a good place to start assuming they all use it anyway.

zonkmachine commented:
> `FileDialog` is a good place to start assuming they all use it anyway.

Yes, they all use `FileDialog` or `VersionedSaveDialog` which is derived from `FileDialog` .
I looked at this before and my answer is the same. I don't know how to do this. I'm not experienced enough with working with Classes. There is a function in there `FileDialog::clearSelection()` which at least gives a hint at how to extend it but it isn't enough.

tresf commented:
> I looked at this before and my answer is the same. I don't know how to do this. I'm not experienced enough with working with Classes.

This time specifically I was speaking about a single, static function to reduce the redundancy in code, e.g. a helper to make a centralized location to test `if (fileExists) { ... }` etc.

Umcaruje commented:
@zonkmachine what's the status on this? Can this still be done for 1.2 or should we move the milestone?

zonkmachine commented:
> what's the status on this?

The status is that the code works and is up to date but adds some redundancy and I don't know how to remedy that.

zonkmachine commented:
> @zonkmachine what's the status on this?

Done! Maybe test it a bit first though... :wink:

zonkmachine commented:
@tresf I've moved the redundant code to a function in versionedFileDialog. Tested and I'm happy with it. Merge?

tresf commented:
@zonkmachine I'll have to defer the decision to merge to someone else.  The squash erased the context of my original questions and I don't have time to review thoroughly.

If you are happy with it please feel free to merge on your own.

zonkmachine commented:
> The squash erased the context of my original questions

:trollface:     A bit of an oops there... I'll merge this then.
